src/library/basetrackcache.cpp

```diff
@@ -180,7 +175,27 @@ void BaseTrackCache::replaceRecentTrack(TrackId trackId, TrackPointer pTrack) co
     if (m_recentTrackId.isValid()) {
         if (pTrack) {
             DEBUG_ASSERT(m_recentTrackId == pTrack->getId());
-            m_recentTrackPtr = std::move(pTrack);
+
+            // NOTE(uklotzde, 2018-02-07, Fedora 27, GCC 7.3.1, optimize=native (-O3), Core i5-6440HQ)
+            // Using the move assignment operator here will sooner or later
+            // store a nullptr in m_recentTrackPtr even if both m_recentTrackPtr
+            // and pTrack contain a valid but different pointer before the
+            // assignment and the custom deleter is invoked on m_recentTrackPtr
+            // during the assignment. The issue does not occur when either
+            // changing the optimization level from 'native' to 'none' or
+            // when replacing the move assignment with a swap operation (see below).
+            //
+            // Fucked up by the optimizer:
+            // m_recentTrackPtr = std::move(pTrack);
```

Comment:
Ups.  We have many statement like this in the code.
Is the a bug report we can reference? Which places may also be effected?

Reply:
No. I wasn't able and did not have the time to isolate this issue for a proper bug report.

Reply:
What means sooner or later? Did this happen with the plain std::shared_ptr or the inherited class or both?

Reply:
As described, after the assignment the left hand side argument is null, even if the right hand side argument was non-null before the assignment. This is impossible and cannot be caused by a race condition if we assume (and we really should) that std::shared_ptr is thread-safe! It happens with the plain std::shared_ptr, but only once in a while. I was not able to create a simple test case that reproduces the crash.

The optimizer might be confused by the cascaded invocation of const member functions that in the end modify some mutable members. Just an idea, but I cannot proof it.

Reply:
Do I need to mention how many hours I spend to discover this subtle bug? I'm really annoyed about C++ now, totally frustrating.